README.md
# Transitive\Web

[![Latest Stable Version](https://poser.pugx.org/transitive/web/v/stable?format=flat-square)](https://packagist.org/packages/transitive/web)
[![License](https://poser.pugx.org/transitive/web/license?format=flat-square)](https://packagist.org/packages/transitive/web)
[![Build Status](https://travis-ci.org/RobinDumontChaponet/TransitiveWeb.svg?branch=master)](https://travis-ci.org/RobinDumontChaponet/TransitiveWeb)
[![Coverage Status](https://coveralls.io/repos/github/RobinDumontChaponet/TransitiveWeb/badge.svg)](https://coveralls.io/github/RobinDumontChaponet/TransitiveWeb)

## Installation

```sh
composer require transitive/web
```

## Basic Usage

```php
<?php

use Transitive\Web;
use Transitive\Routing;

require __DIR__.'/../vendor/autoload.php';

$front = new Web\WebFront();

$front->addRouter(new Routing\PathRouter(dirname(dirname(__FILE__)).'/presenters', dirname(dirname(__FILE__)).'/views'));

$front->execute(@$_GET['request'] ?? 'index');

echo $front;

```

## License

The MIT License (MIT)

Copyright (c) 2016 Robin Dumont-Chaponet

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is (try to) based on [Keep a Changelog](http://keepachangelog.com/) 
and this project adheres to [Semantic Versioning](http://semver.org/).

## 2.2.0
Namespaces have been refactored again…

### Now : 
- Transitive\Core are base classes (package transitive/core)
- Transitive\Simple are basic view and front classes (package transitive/core)
- Transitive\Routing are routing classes (package transitive/routing)
- Transitive\Web are web view and front classes (package transitive/web)

### Coming later :
- Transitive\BulletinBoardSystem
- Transitive\CommandLine




src/View.php
<?php

namespace Transitive\Web;

use Transitive\Core;
use Transitive\Simple;

class View extends Simple\View implements Core\View
{
    /**
     * styles tags and linked scripts.
     *
     * @var array
     */
    public $styles;

    /**
     * scripts tags and linked scripts.
     *
     * @var array
     */
    public $scripts;

    /**
     * metas tags.
     *
     * @var array
     */
    public $metas;

    public function __construct()
    {
        $this->styles = array();
        $this->scripts = array();
        $this->metas = array();
        $this->title = '';
        $this->content = null;
    }

    /**
     * @param string $prefix
     * @param string $separator
     * @param string $endSeparator
     */
    public function getTitle(string $prefix = '', string $separator = ' | ', string $sufix = ''): string
    {
        return parent::getTitle('<title>'.$prefix, $separator, $sufix.'</title>');
    }

    /*
     * @return Core\ViewResource
     */
    public function getHeadValue(): Core\ViewResource
    {
        return new Core\ViewResource(array(
            'metas' => $this->getMetasValue(),
            'title' => $this->getTitleValue(),
            'scripts' => $this->getScriptsValue(),
            'styles' => $this->getStylesValue(),
        ), 'asArray');
    }

    /*
     * @return string
     */
    public function getHead(): string
    {
        return '<head><meta charset="UTF-8">'
               .$this->getHeadValue()->asString()
               .'</head>';
    }

    public function __debugInfo()
    {
        return array(
            'title' => $this->title,
            'metas' => $this->metas,
            'scripts' => $this->scripts,
            'styles' => $this->styles,
            'content' => $this->getContent(),
            'data' => $this->getData(),
        );
    }

    /**
     * @param string $rawTag
     */
    public function addRawMetaTag(string $rawTag): void
    {
        $this->metas[] = array(
            'raw' => $rawTag,
        );
    }

    /**
     * @param string $name
     * @param string $content
     */
    public function addMetaTag(string $name, string $content = ''): void
    {
        $this->metas[] = array(
            'name' => $name,
            'content' => $content,
        );
    }

    /**
     * @return array
     */
    public function getMetasValue(): array
    {
        return $this->metas;
    }

    /*
     * @return string
     */
    public function getMetas(): string
    {
        $str = '';

        if(isset($this->metas))
            foreach($this->metas as $meta)
                if(isset($meta['name']))
                    $str .= '<meta name="'.$meta['name'].'" content="'.$meta['content'].'">';
                else
                    $str .= $meta['raw'];

        return $str;
    }

    /**
     * @param string $content
     * @param string $type
     */
    public function addStyle(string $content, string $type = 'text/css'): void
    {
        $this->styles[] = array(
            'type' => $type,
            'content' => $content,
        );
    }

    /**
     * @param string $content
     * @param string $type
     */
    public function addScript(string $content, string $type = 'text/javascript'): void
    {
        $this->scripts[] = array(
            'type' => $type,
            'content' => $content,
        );
    }

    /**
     * @param string $href
     * @param string $type
     * @param bool   $defer
     * @param bool   $cacheBust
     * @param string $rel
     */
    public function linkStyleSheet(string $href, string $type = 'text/css', bool $defer = false, bool $cacheBust = true, string $rel = 'stylesheet'): void
    {
        if($cacheBust)
            $href = self::cacheBust($href);
        $this->styles[] = array(
            'href' => $href,
            'type' => $type,
            'defer' => $defer,
            'rel' => $rel,
        );
    }

    /**
     * @param string $href
     * @param string $type      = 'text/javascript'
     * @param bool   $defer     = false
     * @param bool   $cacheBust = true
     */
    public function linkScript(string $href, string $type = 'text/javascript', bool $defer = false, bool $cacheBust = true): void
    {
        if($cacheBust)
            $href = self::cacheBust($href);
        $this->scripts[] = array(
            'href' => $href,
            'type' => $type,
            'defer' => $defer,
        );
    }

    /**
     * @param string $rawTag
     */
    public function addRawStyleTag(string $rawTag): void
    {
        $this->styles[] = array(
            'raw' => $rawTag,
        );
    }

    /**
     * @param string $rawTag
     */
    public function addRawScriptTag(string $rawTag): void
    {
        $this->scripts[] = array(
            'raw' => $rawTag,
        );
    }

    /**
     * @param string $filepath
     * @param string $type      = 'text/css'
     * @param bool   $cacheBust = false
     *
     * @return bool
     */
    public function importStyleSheet(string $filepath, string $type = 'text/css', bool $cacheBust = false): bool
    {
        if(!is_file($filepath)) {
            trigger_error('file "'.$filepath.'" failed for import, ressource not found or not a file', E_USER_NOTICE);

            return false;
        }
        if(!is_readable($filepath)) {
            trigger_error('file "'.$filepath.'" failed for import, ressource is not readable', E_USER_NOTICE);

            return false;
        }

        if($cacheBust)
            $filepath = self::cacheBust($filepath);
        $this->addStyle(self::_getIncludeContents($filepath), $type);

        return true;
    }

    /**
     * @param string $filepath
     * @param string $type      = 'text/javascript'
     * @param bool   $cacheBust = false
     *
     * @return bool
     */
    public function importScript(string $filepath, string $type = 'text/javascript', bool $cacheBust = false): bool
    {
        if(!is_file($filepath)) {
            trigger_error('file "'.$filepath.'" failed for import, ressource not found or not a file', E_USER_NOTICE);

            return false;
        }
        if(!is_readable($filepath)) {
            trigger_error('file "'.$filepath.'" failed for import, ressource is not readable', E_USER_NOTICE);

            return false;
        }

        if($cacheBust)
            $filepath = self::cacheBust($filepath);
        $this->addScript(self::_getIncludeContents($filepath), $type);

        return true;
    }

    /*
     * @return string
     */
    public function getStyles(): string
    {
        $str = '';

        if(isset($this->styles))
            foreach($this->styles as $style)
                if(isset($style['content']))
                    $str .= '<style type="'.$style['type'].'">'.$style['content'].'</style>';
                elseif(isset($style['href']))
                    $str .= '<link rel="'.$style['rel'].'" type="'.$style['type'].'" href="'.$style['href'].'"'.(($style['defer']) ? ' defer async' : '').' />';
                elseif(isset($style['raw']))
                    $str .= $style['raw'];

        return $str;
    }

    /*
     * @return string
     */
    public function getStylesContent(): string
    {
        $content = '';
        if(isset($this->styles))
            foreach($this->styles as $style)
                if(isset($style['content']))
                    $content .= $style['content'].PHP_EOL;

        return $content;
    }

    /*
     * @return string
     */
    public function getScripts(): string
    {
        $str = '';

        if(isset($this->scripts))
            foreach($this->scripts as $script)
                if(isset($script['content']))
                    $str .= '<script type="'.$script['type'].'">'.$script['content'].'</script>';
                elseif(isset($script['href']))
                    $str .= '<script type="'.$script['type'].'" src="'.$script['href'].'"'.(($script['defer']) ? ' defer async' : '').'></script>';
                elseif(isset($script['raw']))
                    $str .= $script['raw'];

        return $str;
    }

    /*
     * @return string
     */
    public function getScriptsContent(): string
    {
        $content = '';
        if(isset($this->scripts))
            foreach($this->scripts as $script)
                if(isset($script['content']))
                    $content .= $script['content'];

        return $content;
    }

    /**
     * @return array
     */
    public function getScriptsValue(): array
    {
        return $this->scripts;
    }

    /**
     * @return array
     */
    public function getStylesValue(): array
    {
        return $this->styles;
    }

    /*
     * @param string url
     * @param int delay = 0
     * @param int code = 303
     * @return bool
     */
    public function redirect(string $url, int $delay = 0, int $code = 303): bool
    {
        $this->addRawMetaTag('<meta http-equiv="refresh" content="'.$delay.'; url='.$url.'">');

        if(!headers_sent()) {
            http_response_code($code);
            $_SERVER['REDIRECT_STATUS'] = $code;
            if($delay <= 0)
                header('Location: '.$url, true, $code);
            else
                header('Refresh:'.$delay.'; url='.$url, true, $code);

            return true;
        }

        return false;
    }
}


src/WebFront.php
<?php

namespace Transitive\Front;

use Transitive\Core;
use Transitive\Routing;

function getBestSupportedMimeType($mimeTypes = null) {
    // Values will be stored in this array
    $acceptTypes = array();
    // divide it into parts in the place of a ","
    $accept = explode(',', strtolower(str_replace(' ', '', $_SERVER['HTTP_ACCEPT'])));
    foreach ($accept as $a) {
        // the default quality is 1.
        $q = 1;
        // check if there is a different quality
        if (strpos($a, ';q=')) {
            // divide "mime/type;q=X" into two parts: "mime/type" i "X"
            list($a, $q) = explode(';q=', $a);
        }
        // mime-type $a is accepted with the quality $q
        // WARNING: $q == 0 means, that mime-type isn’t supported!
        $acceptTypes[$a] = $q;
    }
    arsort($acceptTypes);
    // if no parameter was passed, just return parsed data
    if (!$mimeTypes) return $acceptTypes;
    $mimeTypes = array_map('strtolower', (array) $mimeTypes);
    // let’s check our supported types:
    foreach ($acceptTypes as $mime => $q) {
       if ($q && in_array($mime, $mimeTypes)) return $mime;
    }
    // no mime-type found
    return null;
}

/**
 * WebFront class.
 *
 * @extends BasicFront
 * @implements FrontController
 */
class WebFront extends BasicFront implements FrontController
{
    private $httpErrorRoute;
    private static $defaultHttpErrorRoute;

    private $contentType;
    public static $mimeTypes = array(
        'application/xhtml+xml', 'text/html',
        'application/json', 'application/xml',
        'application/vnd.transitive.content+xhtml', 'application/vnd.transitive.content+html',
        'application/vnd.transitive.content+css', 'application/vnd.transitive.content+javascript',
        'application/vnd.transitive.content+json', 'application/vnd.transitive.content+xml', 'application/vnd.transitive.content+yaml',
        'application/vnd.transitive.head+json', 'application/vnd.transitive.head+xml', 'application/vnd.head+yaml',
        'application/vnd.transitive.document+json', 'application/vnd.transitive.document+xml', 'application/vnd.transitive.document+yaml',
        'text/plain',
    );

    public const defaultViewClassName = '\Transitive\Core\WebView';

    public function __construct()
    {
        $this->contentType = getBestSupportedMimeType(self::$mimeTypes);
        $this->obClean = true;
        $this->obContent = '';

        $this->layout = new Routing\Route(new Core\Presenter(), new Core\BasicView());

        $this->setLayoutContent(function ($data) { ?><!DOCTYPE html>
<html>
<head>
<meta charset="UTF-8">
<?= $data['view']->getMetas(); ?>
<?= $data['view']->getTitle('Default layout'); ?>
<?= $data['view']->getStyles(); ?>
<?= $data['view']->getScripts(); ?>
</head>
<body>
	<?= $data['view']; ?>
</body>
</html><?php
        });
    }

    /*
     * @todo remove this ?
     */
    public function getContentType(): ?string
    {
        return $this->contentType;
    }

    protected function _getRoute(string $query, string $defaultViewClassName = null): ?Routing\Route
    {
        try {
            return parent::_getRoute($query, $defaultViewClassName);
        } catch(Routing\RoutingException $e) {
            if($e->getCode() > 200) {
                http_response_code($e->getCode());
                $_SERVER['REDIRECT_STATUS'] = $e->getCode();
            }

            return $this->httpErrorRoute ?? self::$defaultHttpErrorRoute ?? null;
        }
    }

    public function execute(string $queryURL = null): ?Routing\Route
    {
        $this->contentType = getBestSupportedMimeType(self::$mimeTypes);
        $this->route = $this->_getRoute($queryURL, self::defaultViewClassName);

        if(isset($this->route)) {
            try {
                $this->obContent = $this->route->execute($this->obClean);
                $this->executed = true;
            } catch(Routing\RoutingException $e) {
                if($e->getCode() > 200) {
                    http_response_code($e->getCode());
                    $_SERVER['REDIRECT_STATUS'] = $e->getCode();
                }
            } catch(Core\BreakFlowException $e) {
                $this->execute($e->getQueryURL());
            }

            if(/* !$this->route->hasView() ||  */($this->route->hasView() && !$this->route->getView()->hasContent())) {
                http_response_code(204);
                $_SERVER['REDIRECT_STATUS'] = 204;
            }
            if(!empty($this->contentType)) {
                header('Content-Type: '.$this->contentType);
                if(!in_array($this->contentType, array('application/xhtml+xml', 'text/html', 'plain/text'))) {
                    header('Expires: '.gmdate('D, d M Y H:i:s').' GMT');
                    header('Cache-Control: public, max-age=60');
                }
            }
            header('Vary: X-Requested-With,Content-Type');

            $content = ['view' => $this->route->getView()];

            $this->layout->getPresenter()->setData($content);
            $this->layout->execute($this->obClean);
        }

        return $this->route;
    }

    /**
     * @codeCoverageIgnore
     */
    public function __debugInfo()
    {
        return [
            'httpErrorRoute' => $this->httpErrorRoute,
            'routers' => $this->routers,
            'route' => $this->route,
            'obClean' => $this->obClean,
            'obContent' => $this->obContent,
            'executed' => $this->executed,
        ];
    }

    /**
     * @codeCoverageIgnore
     */
    public function __toString(): string
    {
        return $this->getContent();
    }

    /**
     * Return processed content from current route.
     *
     * @return string
     *
     * @param string $contentType = null
     */
    public function getContent(string $contentType = null): string
    {
        if(empty($this->route)) {
            http_response_code(404);
            $_SERVER['REDIRECT_STATUS'] = 404;

            return 'No Route';
        }

        if(null == $contentType)
            $contentType = $this->contentType;
        switch($contentType) {
            case 'application/vnd.transitive.document+json':
                return $this->route->getDocument();
            break;
            case 'application/vnd.transitive.document+xml':
                return $this->route->getDocument()->asXML('document');
            break;
            case 'application/vnd.transitive.document+yaml':
                return $this->route->getDocument()->asYAML();
            break;
            case 'application/vnd.transitive.head+json':
                return $this->route->getHead()->asJson();
            break;
            case 'application/vnd.transitive.head+xml':
                return $this->route->getHead()->asXML('head');
            break;
            case 'application/vnd.transitive.head+yaml':
                return $this->route->getHead()->asYAML();
            break;
            case 'application/vnd.transitive.content+xhtml': case 'application/vnd.transitive.content+html':
                return $this->route->getContent();
            break;
            case 'application/vnd.transitive.content+css':
                return $this->route->getView()->getStylesContent();
            break;
            case 'application/vnd.transitive.content+javascript':
                return $this->route->getView()->getScriptsContent();
            break;
            case 'application/vnd.transitive.content+json':
                return $this->route->getContent()->asJson();
            break;
            case 'application/vnd.transitive.content+xml':
                return $this->route->getContent()->asXML('content');
            break;
            case 'application/vnd.transitive.content+yaml':
                return $this->route->getContent()->asYAML();
            break;

            case 'text/plain':
                return $this->getContent()->asString();
            break;

            case 'application/json':
                if($this->route->hasContent('api'))
                    return $this->route->getContent('api')->asJson();
                elseif(404 != http_response_code()) {
                    http_response_code(404);
                    $_SERVER['REDIRECT_STATUS'] = 404;
                }
            break;
            case 'application/xml':
                if($this->route->hasContent('api'))
                    return $this->route->getContent('api')->asXML();
                elseif(404 != http_response_code()) {
                    http_response_code(404);
                    $_SERVER['REDIRECT_STATUS'] = 404;
                }
            break;

            default:
                return $this->layout->getView();
        }
    }

    public static function setDefaultHttpErrorRoute(Routing\Route $route): void
    {
        self::$defaultHttpErrorRoute = $route;
    }

    public function setHttpErrorRoute(Routing\Route $route): void
    {
        $this->httpErrorRoute = $route;
    }
}

WebFront::setDefaultHttpErrorRoute(new Routing\Route(dirname(getcwd()).'/presenters/genericHttpErrorHandler.php', dirname(getcwd()).'/views/genericHttpErrorHandler.php', null, [], WebFront::defaultViewClassName));


src/Front.php
<?php

namespace Transitive\Web;

use Transitive\Core;
use Transitive\Simple;
use Transitive\Routing;

function getBestSupportedMimeType($mimeTypes = null) {
    // Values will be stored in this array
    $acceptTypes = array();
    // divide it into parts in the place of a ","
    $accept = explode(',', strtolower(str_replace(' ', '', $_SERVER['HTTP_ACCEPT'])));
    foreach ($accept as $a) {
        // the default quality is 1.
        $q = 1;
        // check if there is a different quality
        if (strpos($a, ';q=')) {
            // divide "mime/type;q=X" into two parts: "mime/type" i "X"
            list($a, $q) = explode(';q=', $a);
        }
        // mime-type $a is accepted with the quality $q
        // WARNING: $q == 0 means, that mime-type isn’t supported!
        $acceptTypes[$a] = $q;
    }
    arsort($acceptTypes);
    // if no parameter was passed, just return parsed data
    if (!$mimeTypes) return $acceptTypes;
    $mimeTypes = array_map('strtolower', (array) $mimeTypes);
    // let’s check our supported types:
    foreach ($acceptTypes as $mime => $q) {
       if ($q && in_array($mime, $mimeTypes)) return $mime;
    }
    // no mime-type found
    return null;
}

/**
 * WebFront class.
 *
 * @extends Simple\Front
 * @implements Routing\FrontController
 */
class Front extends Simple\Front implements Routing\FrontController
{
    private $httpErrorRoute;
    private static $defaultHttpErrorRoute;

    private $contentType;
    public static $mimeTypes = array(
        'application/xhtml+xml', 'text/html',
        'application/json', 'application/xml',
        'application/vnd.transitive.content+xhtml', 'application/vnd.transitive.content+html',
        'application/vnd.transitive.content+css', 'application/vnd.transitive.content+javascript',
        'application/vnd.transitive.content+json', 'application/vnd.transitive.content+xml', 'application/vnd.transitive.content+yaml',
        'application/vnd.transitive.head+json', 'application/vnd.transitive.head+xml', 'application/vnd.head+yaml',
        'application/vnd.transitive.document+json', 'application/vnd.transitive.document+xml', 'application/vnd.transitive.document+yaml',
        'text/plain',
    );

    public const defaultViewClassName = '\Transitive\Web\View';

    public function __construct()
    {
        $this->contentType = getBestSupportedMimeType(self::$mimeTypes);
        $this->obClean = true;
        $this->obContent = '';

        $this->layout = new Routing\Route(new Core\Presenter(), new Simple\View());

        $this->setLayoutContent(function ($data) { ?><!DOCTYPE html>
<html>
<head>
<meta charset="UTF-8">
<?= $data['view']->getMetas(); ?>
<?= $data['view']->getTitle('Default layout'); ?>
<?= $data['view']->getStyles(); ?>
<?= $data['view']->getScripts(); ?>
</head>
<body>
	<?= $data['view']; ?>
</body>
</html><?php
        });
    }

    /*
     * @todo remove this ?
     */
    public function getContentType(): ?string
    {
        return $this->contentType;
    }

    protected function _getRoute(string $query, string $defaultViewClassName = null): ?Routing\Route
    {
        try {
            return parent::_getRoute($query, $defaultViewClassName);
        } catch(Routing\RoutingException $e) {
            if($e->getCode() > 200) {
                http_response_code($e->getCode());
                $_SERVER['REDIRECT_STATUS'] = $e->getCode();
            }

            return $this->httpErrorRoute ?? self::$defaultHttpErrorRoute ?? null;
        }
    }

    public function execute(string $queryURL = null): ?Routing\Route
    {
        $this->contentType = getBestSupportedMimeType(self::$mimeTypes);
        $this->route = $this->_getRoute($queryURL, self::defaultViewClassName);

        if(isset($this->route)) {
            try {
                $this->obContent = $this->route->execute($this->obClean);
                $this->executed = true;
            } catch(Routing\RoutingException $e) {
                if($e->getCode() > 200) {
                    http_response_code($e->getCode());
                    $_SERVER['REDIRECT_STATUS'] = $e->getCode();
                }
            } catch(Core\BreakFlowException $e) {
                $this->execute($e->getQueryURL());
            }

            if(/* !$this->route->hasView() ||  */($this->route->hasView() && !$this->route->getView()->hasContent())) {
                http_response_code(204);
                $_SERVER['REDIRECT_STATUS'] = 204;
            }
            if(!empty($this->contentType)) {
                header('Content-Type: '.$this->contentType);
                if(!in_array($this->contentType, array('application/xhtml+xml', 'text/html', 'plain/text'))) {
                    header('Expires: '.gmdate('D, d M Y H:i:s').' GMT');
                    header('Cache-Control: public, max-age=60');
                }
            }
            header('Vary: X-Requested-With,Content-Type');

            $content = ['view' => $this->route->getView()];

            $this->layout->getPresenter()->setData($content);
            $this->layout->execute($this->obClean);
        }

        return $this->route;
    }

    /**
     * @codeCoverageIgnore
     */
    public function __debugInfo()
    {
        return [
            'httpErrorRoute' => $this->httpErrorRoute,
            'routers' => $this->routers,
            'route' => $this->route,
            'obClean' => $this->obClean,
            'obContent' => $this->obContent,
            'executed' => $this->executed,
        ];
    }

    /**
     * @codeCoverageIgnore
     */
    public function __toString(): string
    {
        return $this->getContent();
    }

    /**
     * Return processed content from current route.
     *
     * @return string
     *
     * @param string $contentType = null
     */
    public function getContent(string $contentType = null): string
    {
        if(empty($this->route)) {
            http_response_code(404);
            $_SERVER['REDIRECT_STATUS'] = 404;

            return 'No Route';
        }

        if(null == $contentType)
            $contentType = $this->contentType;
        switch($contentType) {
            case 'application/vnd.transitive.document+json':
                return $this->route->getDocument();
            break;
            case 'application/vnd.transitive.document+xml':
                return $this->route->getDocument()->asXML('document');
            break;
            case 'application/vnd.transitive.document+yaml':
                return $this->route->getDocument()->asYAML();
            break;
            case 'application/vnd.transitive.head+json':
                return $this->route->getHead()->asJson();
            break;
            case 'application/vnd.transitive.head+xml':
                return $this->route->getHead()->asXML('head');
            break;
            case 'application/vnd.transitive.head+yaml':
                return $this->route->getHead()->asYAML();
            break;
            case 'application/vnd.transitive.content+xhtml': case 'application/vnd.transitive.content+html':
                return $this->route->getContent();
            break;
            case 'application/vnd.transitive.content+css':
                return $this->route->getView()->getStylesContent();
            break;
            case 'application/vnd.transitive.content+javascript':
                return $this->route->getView()->getScriptsContent();
            break;
            case 'application/vnd.transitive.content+json':
                return $this->route->getContent()->asJson();
            break;
            case 'application/vnd.transitive.content+xml':
                return $this->route->getContent()->asXML('content');
            break;
            case 'application/vnd.transitive.content+yaml':
                return $this->route->getContent()->asYAML();
            break;

            case 'text/plain':
                return $this->getContent()->asString();
            break;

            case 'application/json':
                if($this->route->hasContent('api'))
                    return $this->route->getContent('api')->asJson();
                elseif(404 != http_response_code()) {
                    http_response_code(404);
                    $_SERVER['REDIRECT_STATUS'] = 404;
                }
            break;
            case 'application/xml':
                if($this->route->hasContent('api'))
                    return $this->route->getContent('api')->asXML();
                elseif(404 != http_response_code()) {
                    http_response_code(404);
                    $_SERVER['REDIRECT_STATUS'] = 404;
                }
            break;

            default:
                return $this->layout->getView();
        }
    }

    public static function setDefaultHttpErrorRoute(Routing\Route $route): void
    {
        self::$defaultHttpErrorRoute = $route;
    }

    public function setHttpErrorRoute(Routing\Route $route): void
    {
        $this->httpErrorRoute = $route;
    }
}

Front::setDefaultHttpErrorRoute(new Routing\Route(dirname(getcwd()).'/presenters/genericHttpErrorHandler.php', dirname(getcwd()).'/views/genericHttpErrorHandler.php', null, [], Front::defaultViewClassName));
